power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -598,10 +600,166 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
+        }
+
+        void adjust(bool strategy_max = true) {
+            if (get_last_down()) {
+                upper_bound_ = current_;
+            } else {
+                lower_bound_ = current_;
+            }
+            if (lower_bound_ < upper_bound_) {
+                bool const prefer_higher = strategy_max != tap_reverse_;
+                IntS const tap_pos = search(prefer_higher);
+                current_ = tap_pos;
+            }
+        }
+
+        IntS search(bool prefer_higher = true) const {
+            // This logic is used to determin which of the middle points could be of interest
+            // given strategy used in optimization:
+            // Since in BinarySearch we insist on absolute upper and lower bounds, we only need to
+            // find the corresponding mid point. std::midpoint returns always the lower mid point
+            // if the range is of even length. This is why we need to adjust bounds accordingly.
+            // Not because upper bound and lower bound might be reversed, which is not possible.
+            auto const primary_bound = prefer_higher ? upper_bound_ : lower_bound_;
+            auto const secondary_bound = prefer_higher ? lower_bound_ : upper_bound_;
+            return std::midpoint(primary_bound, secondary_bound);
+        }
+
+        IntS lower_bound_{};         // tap position lower bound
+        IntS upper_bound_{};         // tap position upper bound
+        IntS current_{0};            // current tap position
+        bool last_down_{false};      // last direction
+        bool last_check_{false};     // last run checked
+        bool tap_reverse_{false};    // tap range normal or reversed
+        bool inevitable_run_{false}; // inevitable run
+    };
+    std::vector<std::vector<BinarySearch>> binary_search_;
+    struct BinarySearchOptions {
+        bool strategy_max{false};
+        Idx2Du idx_bs{0, 0};
+    };
+    Idx total_iterations{0}; // metric purpose only
+
   public:
     TapPositionOptimizerImpl(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy,
-                             meta_data::MetaData const& meta_data)
-        : meta_data_{&meta_data}, calculate_{std::move(calculator)}, update_{std::move(updater)}, strategy_{strategy} {}
+                             meta_data::MetaData const& meta_data,
+                             std::optional<SearchMethod> tap_search = std::nullopt)
+        : meta_data_{&meta_data}, calculate_{std::move(calculator)}, update_{std::move(updater)}, strategy_{strategy} {
+        auto const is_supported = [&strategy](std::optional<SearchMethod> const& search) {
+            if (!search) {
+                return true;
+            }
+            switch (strategy) {
+            case OptimizerStrategy::any:
+                return search == SearchMethod::linear_search;
+            case OptimizerStrategy::fast_any:
+                return search == SearchMethod::binary_search;
```

Comment:
i don't want you to change everything but it's just for my understanding cause I woke up in the middle of the night thinking about this:

is there a particular reason why we don't use `OptimizerStrategy::any` here and support both `linear_search` and `binary_search`, and that the `fast_any` is just the way we expose `any + binary_search` to the user? from an application perspective, that would look somewhat like

| optimization strategy \ search strategy | linear          | binary           |
| --------------------------------------- | :-------------: | :--------------: |
| any                                     | `any_valid_tap` | `fast_any_tap`   |
| local minimum                           | N/A             | `global_minimum` |
| global minimum                          | N/A             | `local_minimum`  |
| local maximum                           | N/A             | `global_maximum` |
| global maximum                          | N/A             | `local_maximum`  |

that way, all combinations and search strategies are supported from a tap position optimizer point of view, and only on the interface side a subset of those combinations is exposed

Reply:
See our decided way-forward in the Jira ticket. I personally do not have a strong opinion here. One argument against this might be the convolution of concepts.

Reply:
Since this is an internal thing, I guess we can do a follow up PR optimizing the number of internal enums if necessary.

Reply:
sounds good to me